tools/codegen/api/translate.py

```diff
@@ -47,12 +47,21 @@ class UnsatError(RuntimeError):
 #
 # TODO: Don't need full Binding for goals, CType will do
 # TODO: Don't need full Binding for bindings, list of Expr will do
-def translate(bindings: Sequence[Binding], goals: Sequence[Binding], *, method: bool = False) -> List[Expr]:
+def translate(
+    bindings: Sequence[Binding], goals: Sequence[Binding], *,
+    method: bool = False, skip_possibly_redundant_memory_format: bool = False
+) -> List[Expr]:
     # Add all the bindings to the context
     ctx: Dict[CType, str] = {}
     for b in bindings:
         ctx[b.ctype] = b.name
 
+        # handle `const c10::optional<Tensor> &` -> `const Tensor &`
+        t = b.ctype
+        if isinstance(t, ConstRefCType) and isinstance(t.elem, OptionalCType) and \
+           isinstance(t.elem.elem, BaseCType) and t.elem.elem.type == 'Tensor':
+            ctx[ConstRefCType(BaseCType("Tensor", b.name))] = f'({b.name}.has_value() ? *{b.name} : at::Tensor())'
```

Comment:
nit: Today, functionally it doesn't make a difference, but it would be better to put this translation rule inside `solve` itself, so that we are still uniformly doing backward inference. Because the rule here is very simple it can be done with forward and backward, so it's mostly a uniformity thing.

Reply:
(The trouble with forward inference is when you start stuffing the context with tons and tons of possible conversions "just because they might help"; backward lets you be a lot more directed about things. Though it's not that bad of an idea; see Datalog for example :)

Reply:
You know what I changed my mind, unpacking of optional to tensor should be done as forward inference.